pkg/kv/kvclient/rangefeed/rangefeed.go

```diff
@@ -427,54 +433,91 @@ func (f *RangeFeed) processEvents(
-						"received unexpected rangefeed SST event with no OnSSTable handler")
-				}
-				f.onSSTable(ctx, ev.SST, ev.RegisteredSpan)
-			case ev.DeleteRange != nil:
-				if f.onDeleteRange == nil {
-					if f.knobs != nil && f.knobs.IgnoreOnDeleteRangeError {
-						continue
-					}
-					return errors.AssertionFailedf(
-						"received unexpected rangefeed DeleteRange event with no OnDeleteRange handler: %s", ev)
-				}
-				f.onDeleteRange(ctx, ev.DeleteRange)
-			case ev.Metadata != nil:
-				if f.onMetadata == nil {
-					return errors.AssertionFailedf("received unexpected metadata event with no OnMetadata handler")
-				}
-				f.onMetadata(ctx, ev.Metadata)
-			case ev.Error != nil:
-				// Intentionally do nothing, we'll get an error returned from the
-				// call to RangeFeed.
+			if err := f.processEvent(ctx, frontier, ev.RangeFeedEvent, ev.RegisteredSpan); err != nil {
+				return err
 			}
 		case <-ctx.Done():
 			return ctx.Err()
 		}
 	}
 }
+
+func (f *RangeFeed) processEvent(
+	ctx context.Context, frontier span.Frontier, ev *kvpb.RangeFeedEvent, registeredSpan roachpb.Span,
+) error {
+	switch {
+	case ev.Val != nil:
+		f.onValue(ctx, ev.Val)
+	case ev.Checkpoint != nil:
+		ts := ev.Checkpoint.ResolvedTS
+		if f.frontierQuantize != 0 {
+			ts.Logical = 0
+			ts.WallTime -= ts.WallTime % int64(f.frontierQuantize)
+		}
+		advanced, err := frontier.Forward(ev.Checkpoint.Span, ts)
+		if err != nil {
+			return err
+		}
+		if f.onCheckpoint != nil {
+			f.onCheckpoint(ctx, ev.Checkpoint)
+		}
+		if advanced && f.onFrontierAdvance != nil {
+			f.onFrontierAdvance(ctx, frontier.Frontier())
+		}
+		if f.frontierVisitor != nil {
+			f.frontierVisitor(ctx, advanced, frontier)
+		}
+	case ev.SST != nil:
+		if f.onSSTable == nil {
+			return errors.AssertionFailedf(
+				"received unexpected rangefeed SST event with no OnSSTable handler")
+		}
+		f.onSSTable(ctx, ev.SST, registeredSpan)
+	case ev.DeleteRange != nil:
+		if f.onDeleteRange == nil {
+			if f.knobs != nil && f.knobs.IgnoreOnDeleteRangeError {
+				return nil
+			}
+			return errors.AssertionFailedf(
+				"received unexpected rangefeed DeleteRange event with no OnDeleteRange handler: %s", ev)
+		}
+		f.onDeleteRange(ctx, ev.DeleteRange)
+	case ev.Metadata != nil:
+		if f.onMetadata == nil {
+			return errors.AssertionFailedf("received unexpected metadata event with no OnMetadata handler")
+		}
+		f.onMetadata(ctx, ev.Metadata)
+	case ev.Error != nil:
+		// Intentionally do nothing, we'll get an error returned from the
+		// call to RangeFeed.
+	case ev.BulkEvents != nil:
+		if f.onValues != nil {
+			allValues := true
+			buf := make([]kv.KeyValue, len(ev.BulkEvents.Events))
+			for i := range ev.BulkEvents.Events {
+				if ev.BulkEvents.Events[i].Val != nil {
+					buf[i] = kv.KeyValue{
+						Key:   ev.BulkEvents.Events[i].Val.Key,
+						Value: &ev.BulkEvents.Events[i].Val.Value,
+					}
+				} else {
+					allValues = false
+					break
+				}
+			}
+			if allValues {
```

Comment:
If we wanted we could instead keep track of the last index of a value, flush the prefix of values, then iterate through the rest of the events one by one.